File: Problem2_CousinsInBinaryTree.cpp
// Time Complexity : O(n) where n is number of elements as we are traversing for each element
// Space Complexity : O(n) actually it is 2^h the last row will have worst case scenario n/2 so its O(n)
// Did this code successfully run on Leetcode :yes
// Did this code successfully run on Home Computer :No
// Any problem you faced while coding this : no


// Your code here along with comments explaining your approach


#include<vector>
#include<bits/stdc++.h>
#include<stdlib.h>
#include<cmath>
#include<time.h>
#include<iostream>

using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    TreeNode* x_parent;
    TreeNode* y_parent;
    int x_depth;
    int y_depth;
public:
    bool isCousins(TreeNode* root, int x, int y) {
        dfs(root, NULL, 0,x ,y);
        return x_parent != y_parent && x_depth == y_depth;
        
    }
        
private:
    void dfs(TreeNode* root, TreeNode* parent, int depth, int x, int y){ // we need to check x and y if its present and then check the level of each and their parent in the recusrsive stack
        //base
        if(root==NULL) return;
        
        //logic
        if(root->val==x){
            x_parent=parent;
            x_depth=depth;
        }
        if(root->val==y){
            y_parent=parent;
            y_depth=depth;
        }
        dfs(root->left,root,depth+1,x,y);
        //st.pop()
        dfs(root->right,root,depth+1,x,y);
         
               
    }    
};

// class Solution {

// public:
//     bool isCousins(TreeNode* root, int x, int y) {
        
//         queue<TreeNode*> q; //in BFSwe have a queue
//         q.push(root); //first we must push the node in the queue
        
//         while(!q.empty()){
//             int size=q.size();
//             bool x_found=false;
//             bool y_found=false;
            
//             for(int i=0; i<size; i++){
                
//                 TreeNode* curr=q.front(); //in JAVA this is TreeNode curr=q.poll()
//                 q.pop();
                
//                 if(curr->val==x) x_found=true;
//                 if(curr->val==y) y_found=true;
//                 if(curr->left != NULL && curr->right != NULL){
//                     if(curr->left->val == x && curr->right->val ==y) return false;
//                     if(curr->left->val == y && curr->right->val ==x) return false;
//                 }
//                 if(curr->left != NULL) q.push(curr->left);
//                 if(curr->right != NULL) q.push(curr->right); 
//             }
//             if(x_found== true && y_found==true) return true;
//             if(x_found== true || y_found==true) return false;
            
//         }
//         return false;
        
//     }
        
  
// };

File: Problem_2 Cousins_in_Binary_Tree.py
# DFS Solution:
# // Time Complexity : O(n)
# // Space Complexity : O(n)
# // Did this code successfully run on Leetcode : Yes
# // Any problem you faced while coding this : No
#
#
# // Your code here along with comments explaining your approach


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:

    def __init__(self):
        self.x_parent = None
        self.y_parent = None
        self.x_depth = None
        self.y_depth = None

    def isCousins(self, root: TreeNode, x: int, y: int) -> bool:
        self.dfs(root, None, 0, x, y)
        return self.x_depth == self.y_depth and self.x_parent != self.y_parent

    def dfs(self, root, parent, depth, x, y):
        if root == None:
            return
        if root.val == x:
            self.x_depth = depth
            self.x_parent = parent
        if root.val == y:
            self.y_depth = depth
            self.y_parent = parent
        self.dfs(root.left, root, depth + 1, x, y)
        self.dfs(root.right, root, depth + 1, x, y)

# BFS Solution:
# // Time Complexity : O(n)
# // Space Complexity : O(n)
# // Did this code successfully run on Leetcode : Yes
# // Any problem you faced while coding this : No
#
#
# // Your code here along with comments explaining your approach

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def isCousins(self, root: TreeNode, x: int, y: int) -> bool:
        q = deque()
        q.append(root)
        while q:
            size = len(q)
            x_found = False
            y_found = False
            for i in range(size):
                curr = q.popleft()
                if curr.val == x: x_found = True
                if curr.val == y: y_found = True
                if curr.left != None and curr.right != None:
                    if curr.left.val == x and curr.right.val == y: return False
                    if curr.right.val == x and curr.left.val == y: return False
                if curr.left != None: q.append(curr.left)
                if curr.right != None: q.append(curr.right)
            if x_found and y_found: return True
            if x_found or y_found: return False
        return False
